### DISCUSSION TITLE
Help understanding why `execute.capture` isn't capturing

### DISCUSSION THREAD

**Discussion by Rouneq** [rakshasa/rtorrent#1278] [Q&A] [answered]
I've searched around looking for examples of `execute.capture` on shell variables. To start, I'm trying to set a value of the home directory. I started with an example found [here](https://github.com/rakshasa/rtorrent/issues/660#issuecomment-347016255):

```
method.insert = completion_dirname, simple|private,
"execute.capture = bash, -c, "echo \"$1\" | tr -d '\\n'", completion_dirname, $d.get_finished_dir="
```

and modified it accordingly to

`method.insert = cfg.home, string|const|private,"execute.capture = bash, -c, "echo $HOME""`

Restarting rtorrent gives me "Junk at end of input". I didn't think this was right, but I tried it anyway.  Changed it to

`method.insert = cfg.home, string|const|private,"execute.capture = bash, -c, \"echo $HOME\""`

With this, if I execute an rpc call with `d.multicall2 '' default cfg.home=`, I get

`'execute.capture = bash, -c, "echo $HOME"'` for each torrent.

What am I missing here?

**Comment by kannibalox** [comment, marked as answer]
`string|const|private` means that this command can only ever return a constant string, which is what you're seeing. You'd need to change it back to just `string|private`, but I'd recommend using [`system.env`](https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#term-system-env) instead for this specific use case.

**Reply by Rouneq** [reply]
Hmm. Alright. I still think I'll need `execute.capture` or something isn't copacetic about `system.env`. HOME was just one value I'm after (which I pulled in via the change you mentioned above). I'm also looking for HOSTNAME and it's coming back as an empty string. But `echo $HOSTNAME` gives me the expected value. And making the call via `execute.capture` still just gives me `'execute.capture = bash, -c, "echo $HOSTNAME"'` Thoughts?

**Reply by kannibalox** [reply]
The `$HOSTNAME` from bash is actually a shell variable and not environment variable (similar to how `$RANDOM` also looks like an environment variable but still returns a different value each time), while `$HOME` is typically set as a "real" environment variable on process start. I'd still recommend passing it in as a proper environment variable via your service manager and using `system.env`, but if that's not an option, something like this should work:
```
method.insert = cfg.home,  private|string, (execute.capture,bash,-c,"echo -n HOSTNAME")
```
Note that it runs execute.capture only once at start-up.

**Reply by Rouneq** [reply]
Awesome. That got me really close. I needed `$HOSTNAME`, but otherwise it picked up the value just fine.

I"m good with it capturing upon startup. That seems like expected behavior.

**Reply by Rouneq** [reply]
@kannibalox, if you'll indulge one more line of inquiry... :) Given

```
method.insert = cfg.user, private|string, (cat, (system.env, USER), "")
method.insert = cfg.host,  private|string, (execute.capture,bash,-c,"echo -n $HOSTNAME")
method.insert = cfg.home, private|string, (cat, (system.env, HOME), "")
method.insert = cfg.basedir,  private|string, (cat,(cfg.home),"/private/rtorrent/")

# API
network.scgi.open_local = <basedir>/private/rtorrent/.socket
# network.scgi.open_local = (cat, (cfg.basedir), ".socket")
execute2 = {sh,-c,/usr/bin/php <basedir>/www/<username>.<server>.feralhosting.com/public_html/rutorrent/php/initplugins.php <username> &}
# execute2 = {sh,-c,/usr/bin/php (cat, (cfg.home), "/www/", (cfg.user), ".", (cfg.host), ".feralhosting.com/public_html/rutorrent/php/initplugins.php") (cat, (cfg.user)) &}
```

where

`rpc method.get '' cfg.basedir`

results in

`<basedir>/private/rtorrent/`

If I use the line 

`# network.scgi.open_local = (cat, (cfg.basedir), ".socket")`

rtorrent starts, but attempts to use rpc result in

```
ERROR:pyrosimple.scripts.rtxmlrpc.RtorrentXmlRpc:While calling method.get('', 'cfg.host') via <RTorrentProxy via xml for scgi+unix:///private/rtorrent/.socket>: [Errno 2] No such file or directory
```

It's like `cfg.basedir` isn't actually set (but getting the value indicates it is), thus making me believe I'm not using `cat` correctly here.

Restoring the first line back to the hard-coded path and using the line

```
execute2 = {sh,-c,/usr/bin/php (cat, (cfg.home), "/www/", (cfg.user), ".", (cfg.host), ".feralhosting.com/public_html/rutorrent/php/initplugins.php") (cat, (cfg.user)) &}
```

rtorrent won't start with the message

`rtorrent: Error in option file: ~/.rtorrent.rc:14: Could not find closing '}'.`

What's odd is I use these exact same new lines on another server and it works. I've pulled the lines from both configs and they match up. The only difference between the servers is the one that doesn't work is a new server I set up earlier today and it's not currently processing any torrents.

**Reply by kannibalox** [reply]
> ```
> ERROR:pyrosimple.scripts.rtxmlrpc.RtorrentXmlRpc:While calling method.get('', 'cfg.host') via <RTorrentProxy via xml for scgi+unix:///private/rtorrent/.socket>: [Errno 2] No such file or directory
> ```
You're not doing anything wrong in the rtorrent.rc here, it's just that pyrosimple's connection auto-detection can't handle all the edge cases that could exist in rtorrent.rc. It _should_ be able to handle nested `cat`ed cfg strings, but without replicating the full parsing engine it'll never be perfect. You can use the `-U` flag to set a connnection string manually, or set it in the config file: https://kannibalox.github.io/pyrosimple/configuration/#scgi_url-pyro_scgi_url

> ```
> execute2 = {sh,-c,/usr/bin/php (cat, (cfg.home), "/www/", (cfg.user), ".", (cfg.host), ".feralhosting.com/public_html/rutorrent/php/initplugins.php") (cat, (cfg.user)) &}
> ```
> rtorrent won't start with the message
> `rtorrent: Error in option file: ~/.rtorrent.rc:14: Could not find closing '}'.`

This is another nesting issue, and the solution is probably more `cfg.` commands so that you can avoid having to use `cat` inside of the `execute2` statement.

**Comment by zeidoo** [comment]
I have a similar problem. I'm not really sure why it's so difficult to read a variable from a file in the rtorrent configs/scripting. I just want to read my dynamic port from a file.

After struggling for the better part of the day trying various combinations, I found that:
- using `more` will capture the file location header as for some reason calling it from rtorrent makes it think that you are more multiple files. The variable will contain something like ::::::::: file/path :::::::::: \n content.
- using cat will add a new line at the end, so the variable appears empty,

I finally managed to get the file content in a variable using the `tr -d` part below.
```
method.insert = test.port, private|const|simple, (execute.capture,"bash",-c,(cat,"cat", " /tmp/wg-tcp-port", " | tr -d '\\n'"))
print=(test.port)
```

But now I get this for any other line below the `method.insert` one.
"Failed to parse command line option: Error in option file: /root/rtorrent.rc:84: Invalid start of command name."

The execution log shows nothing of use:
```
---
sh -c echo >/var/lib/rtorrent/session/rtorrent.pid 2797
---

--- Success ---

---
bash -c cat /tmp/wg-tcp-port | tr -d '\n'
---
Captured output:
41926
--- Success ---
```

Could you kindly give some pointers to solve this?

**Comment by rakshasa** [comment]
```
diff --git a/src/rpc/parse_commands.cc b/src/rpc/parse_commands.cc
index 964c52a..77eae42 100644
--- a/src/rpc/parse_commands.cc
+++ b/src/rpc/parse_commands.cc
@@ -73,7 +73,7 @@ parse_command_execute(target_type target, torrent::Object* object) {
 inline const char*
 parse_command_name(const char* first, const char* last, char* dest_first, char* dest_last) {
   if (first == last || !std::isalpha(*first))
-    throw torrent::input_error("Invalid start of command name.");
+    throw torrent::input_error("Invalid start of command name: '" + std::string(first, last) + "'");
 
   last = first + std::min(std::distance(first, last), std::distance(dest_first, dest_last) - 1);
 
```

Try the above patch to see the command that causes the error.
